### ISSUE TITLE
Bug: X86-only memory leak

### ISSUE THREAD

**Issue by Gero5** [mpogue2/SquareDesk#945]
v 1.0.6 X
I had the app run over night. iMac went to sleep. next morning it was completely dead. there is a high propability that SquareDesk caused the problem.

The Apple error report follows, if that is helpful in any way.

```
panic(cpu 2 caller 0xffffff800a25c89b): userspace watchdog timeout: no successful checkins from configd (2 induced crashes) in 180 seconds
service returned not alive with context : unresponsive dispatch queue(s): com.apple.main-thread 
configd has not exited since first loaded
service: logd, total successful checkins in 158123 seconds: 15687, last successful checkin: 0 seconds ago
service: WindowServer, total successful checkins in 158082 seconds: 15683, last successful checkin: 0 seconds ago
service: opendirectoryd, total successful checkins in 158123 seconds: 15686, last successful checkin: 0 seconds ago
service: configd (2 induced crashes), total successful checkins in 158123 seconds: 15668, last successful checkin: 180 seconds ago

Panicked task 0xffffff868db47848: 6 threads: pid 121: watchdogd
Backtrace (CPU 2), panicked thread: 0xffffff9026b710c8, Frame : Return Address
0xffffffc481e57510 : 0xffffff800763aca1 
0xffffffc481e57560 : 0xffffff80077993c0 
0xffffffc481e575a0 : 0xffffff8007788b0c 
0xffffffc481e57620 : 0xffffff80075d7971 
0xffffffc481e57640 : 0xffffff800763af8d 
0xffffffc481e57730 : 0xffffff800763a633 
0xffffffc481e57790 : 0xffffff8007dd96de 
0xffffffc481e57880 : 0xffffff800a25c89b 
0xffffffc481e57890 : 0xffffff800a25c364 
0xffffffc481e578b0 : 0xffffff800a25b42b 
0xffffffc481e579e0 : 0xffffff8007d4ee56 
0xffffffc481e57a10 : 0xffffff8007d4f1cd 
0xffffffc481e57b80 : 0xffffff80077445df 
0xffffffc481e57ca0 : 0xffffff80076138a0 
0xffffffc481e57d60 : 0xffffff800762b6b8 
0xffffffc481e57dd0 : 0xffffff800762bd28 
0xffffffc481e57ef0 : 0xffffff800776e102 
0xffffffc481e57fa0 : 0xffffff80075d7dd6 
      Kernel Extensions in backtrace:
         com.apple.driver.watchdog(1.0)[F1797B78-7645-3FC6-AD29-624DD0826DA0]@0xffffff800a25a000->0xffffff800a25cfff

Process name corresponding to current thread (0xffffff9026b710c8): watchdogd

Mac OS version:
23B81

Kernel version:
Darwin Kernel Version 23.1.0: Mon Oct  9 21:27:27 PDT 2023; root:xnu-10002.41.9~6/RELEASE_X86_64
Kernel UUID: 7C8FE931-57C6-38AD-9EBB-CC5572293AE6
roots installed: 0
KernelCache slide: 0x0000000007200000
KernelCache base:  0xffffff8007400000
Kernel slide:      0x00000000072e0000
Kernel text base:  0xffffff80074e0000
__HIB  text base: 0xffffff8007300000
System model name: iMac19,1 (Mac-AA95B1DDAB278B95)
System shutdown begun: NO
Panic diags file available: YES (0x0)
Hibernation exit count: 0

System uptime in nanoseconds: 158123186939053
Last Sleep:           absolute           base_tsc          base_nano
  Uptime  : 0x00008fcfec636784
  Sleep   : 0x0000041d654889d8 0x00000000cf4dfb98 0x0000041194559899
  Wake    : 0x0000041d6c6cccda 0x00000000ceab36c6 0x0000041d6a4c231b
Compressor Info: 25% of compressed pages limit (OK) and 12% of segments limit (OK) with 4 swapfiles and OK swap space
Zone info:
  Zone map: 0xffffff8025fb4000 - 0xffffffa025fb4000
  . PGZ   : 0xffffff8025fb4000 - 0xffffff8026fb5000
  . VM    : 0xffffff8026fb5000 - 0xffffff84f3a1b000
  . RO    : 0xffffff84f3a1b000 - 0xffffff868d2e7000
  . GEN0  : 0xffffff868d2e7000 - 0xffffff8b59d4d000
  . GEN1  : 0xffffff8b59d4d000 - 0xffffff90267b3000
  . GEN2  : 0xffffff90267b3000 - 0xffffff94f3219000
  . GEN3  : 0xffffff94f3219000 - 0xffffff99bfc80000
  . DATA  : 0xffffff99bfc80000 - 0xffffffa025fb4000
  Metadata: 0xffffffcf64c20000 - 0xffffffcf84c20000
  Bitmaps : 0xffffffcf84c20000 - 0xffffffcf86420000
  Extra   : 0 - 0

last started kext at 120910931806178: com.paragon-software.lvm-for-mac	1 (addr 0xffffff7f9d4e4000, size 81920)
last stopped kext at 121161188517346: com.paragon-software.lvm-for-mac	1 (addr 0xffffff7f9d4e4000, size 81920)
loaded kexts:
com.paragon-software.filesystems.extfs	237.6.11
@filesystems.smbfs	5.0
>!ATopCaseHIDEventDriver	7400.26
>!APlatformEnabler	2.7.0d0
>AGPM	135
>X86PlatformShim	1.0.0
>AudioAUUC	1.70
>!AUpstreamUserClient	3.6.11
@kext.AMDFramebuffer	5.0.1
>!AHDA	500.3
@filesystems.autofs	3.0
@kext.AMDRadeonX4000	5.0.1
@kext.AMDRadeonServiceManager	5.0.1
>!A!IKBLGraphics	22.0.1
>!AGraphicsDevicePolicy	8.1.9
@AGDCPluginDisplayMetrics	8.1.9
@kext.AMD9500!C	5.0.1
>pmtelemetry	1
@filesystems.nfs	1
|IOUserEthernet	1.0.1
>usb.!UUserHCI	1
>!A!ICFLGraphicsFramebuffer	22.0.1
>!ASunrise	1
>!AHV	1
>!A!IPCHPMC	2.0.1
>!A!ISlowAdaptiveClocking	4.0.0
>!ASMCLMU	7400.2
>!AGFXHDA	300.1
>!AMCCSControl	1.17
>!AThunderboltIP	4.0.3
>!AWWANSupport	4.0.0
>BCMWLANFirmware4388.Hashstore	1
>BCMWLANFirmware4387.Hashstore	1
>BCMWLANFirmware4378.Hashstore	1
>BCMWLANFirmware4377.Hashstore	1
>BCMWLANFirmware4364.Hashstore	1
>BCMWLANFirmware4355.Hashstore	1
>!AFileSystemDriver	3.0.1
@filesystems.tmpfs	1
@filesystems.lifs	1
@filesystems.hfs.kext	650.0.2
@BootCache	40
@!AFSCompression.!AFSCompressionTypeZlib	1.0.0
@!AFSCompression.!AFSCompressionTypeDataless	1.0.0d1
@filesystems.apfs	2235.41.1
>!ASDXC	3.4.3
|!ABCM5701Ethernet	11.0.0
>!ABCMWLANBusInterfacePCIeMac	1
>!AAHCIPort	380
@private.KextAudit	1.0
>!AACPIButtons	6.1
>!ARTC	2.0.1
>!ASMBIOS	2.1
>!AACPIEC	6.1
>!AAPIC	1.7
@!ASystemPolicy	2.0.0
@nke.applicationfirewall	404
|IOKitRegistryCompatibility	1
|EndpointSecurity	1
@Dont_Steal_Mac_OS_X	7.0.0
>!ADiskImages2	273
@kec.!AEncryptedArchive	1
>!AActuatorDriver	7400.42
>!AHIDKeyboard	7400.2
>!AMultitouchDriver	7400.42
>!AInputDeviceSupport	7410.1
>!AHS!BDriver	7400.26
>IO!BHIDDriver	9.0.0
>DspFuncLib	500.3
@kext.OSvKernDSPLib	530
@kext.AMDRadeonX4400HWLibs	1.0
@kext.AMDRadeonX4000HWServices	5.0.1
>!ASMBusPCI	1.0.14d1
>!AGraphicsControl	8.1.9
>!A!ILpssUARTv1	3.0.60
>!A!ILpssUARTCommon	3.0.60
>!AOnboardSerial	1.0
@plugin.IOAVBDiscoveryPlugin	1200.18
@kext.AMDSupport	5.0.1
@kext.triggers	1.0
|IOAccelerator!F2	481.0.1
>IOHIDPowerSource	1
>!ACallbackPowerSource	1
>!ARSMChannel	1
|IORSM!F	1
>!AIPAppender	1.0
@!AGPUWrangler	8.1.9
@!AGraphicsDeviceControl	8.1.9
|IOSlowAdaptiveClocking!F	1.0.0
>!AHDA!C	500.3
|IOHDA!F	500.3
|IONDRVSupport	598
|IOGraphics!F	598
>!ASMBus!C	1.0.18d1
>!UAudio	610.3
|IOAudio!F	500.4
@vecLib.kext	1.2.0
>!AAudioClockLibs	300.1
|IOAVB!F	1200.18
@plugin.IOgPTPPlugin	1200.91
>X86PlatformPlugin	1.0.0
>IOPlatformPlugin!F	6.0.0d8
>!AThunderboltDPOutAdapter	8.5.1
>driverkit.serial	6.0.0
>usb.cdc	5.0.0
>usb.networking	5.0.0
>usb.!UHostCompositeDevice	1.2
>!AThunderboltPCIDownAdapter	4.1.1
>!AThunderboltDPInAdapter	8.5.1
>!AThunderboltDPAdapter!F	8.5.1
>!AHPM	3.4.4
>!A!ILpssI2C!C	3.0.60
>!A!ILpssI2C	3.0.60
>!A!ILpssDmac	3.0.60
>!ABSDKextStarter	3
|IOSurface	352.0.3
@filesystems.hfs.encodings.kext	1
>!ASyntheticGame!C	11.1.9
>!AXsanScheme	3
>!AThunderboltNHI	7.2.81
|IOThunderbolt!F	9.3.3
|IONVMe!F	2.1.0
>!ABCMWLANCoreMac	1.0.0
|IO80211!F	1200.13.0
>IOImageLoader	1.0.0
>!AOLYHALMac	1
|IOSerial!F	11
>corecapture	1.0.4
>!AEFINVRAM	2.1
>usb.!UHostPacketFilter	1.0
|IOUSB!F	900.4.2
|IOAHCI!F	304
>usb.!UXHCIPCI	1.2
>usb.!UXHCI	1.2
>!ASMCRTC	1.0
>!AEFIRuntime	2.1
|IOSMBus!F	1.1
|IOHID!F	2.0.0
|IOTimeSync!F	1200.91
|IOSkywalk!F	1.0
>mDNSOffloadUserClient	1.0.1b8
|IONetworking!F	3.4
>DiskImages	493.0.0
|IO!B!F	9.0.0
|IOReport!F	47
$quarantine	4
$sandbox	300.0
@kext.!AMatch	1.0.0d1
>!ASSE	1.0
>!ALockdownMode	1
>!AKeyStore	2
>!UTDM	556
|IOUSBMass!SDriver	243
|IOSCSIBlockCommandsDevice	492
|IOSCSIArchitectureModel!F	492
>!AFDEKeyStore	28.30
>!AEffaceable!S	1.0
|IO!S!F	2.1
>!ACyrus	1
>!ACredentialManager	1.0
>KernelRelayHost	1
|IOUSBHost!F	1.2
>!UHostMergeProperties	1.2
>usb.!UCommon	1.0
>!AMobileFileIntegrity	1.0.5
$!AImage4	5.0.0
@kext.CoreTrust	1
|CoreAnalytics!F	1
>!ABusPower!C	1.0
>!ASEPManager	1.0.1
>IOSlaveProcessor	1
>!AACPIPlatform	6.1
>!ASMC	3.1.9
|IOPCI!F	2.9
|IOACPI!F	1.4
>watchdog	1
@kec.pthread	1
@kec.XrtHostedXnu	1
@kec.Libm	1
@kec.Compression	1.0
@kec.corecrypto	14.0



```

**Comment by mpogue2** [comment]
I'm watching memory here, and I don't see any memory leaks with SquareDesk at all (M1 MacBook Air running Ventura).

I poked around on the internet, and this particular problem may have several causes, including old/bad kernel extensions:
- https://discussions.apple.com/thread/254541933

So, things to check:
- what OS are you running?  Is it up to date?
- Running any kernel extensions?

I'll do a check on my X86 system to watch for memory leaks, too, and I'll post the results here.

**Comment by mpogue2** [comment]
Yes, it is leaking on my X86 laptop but not on my M1 laptop.  About 500KB per second.

**Comment by mpogue2** [comment]
OK, the setLine calls in the svgVUMeter update() seem to be leaking, BUT only on X86.  That's probably internal to the Qt library, which makes it very hard for us to fix.  However, I have put in code that will minimize the calls to setLine when SquareDesk is not playing music (on ALL platforms), and that should reduce the svgVUMeter leakage to a very very small number that should not ever crash the machine overnight.

Commit: 9559d47b1ede67a1e2f74024d132921e9ea34f72

NOTE: there may be other leaks, which I'll have to try to track down one by one.  I was also chasing a setLine leak in the svgClock, but it was much much slower than the svgVUmeter leak.  The svgClock one will be very hard to get rid of, without modifying the Qt library itself.  Or, we could turn off the second hand of the clock to make it go slower (but it's already quite slow, maybe 50kB/sec).

**Comment by mpogue2** [comment]
I think the leaks are mostly fixed now (see #966).  In any case, I don't expect any crashes due to very small leaks, even when run overnight.

So, let's close this one and reopen if we see overnight crashes again (even so, it might not be memory leaks that caused it, we'll have to dive deeper on that if/when it happens again...)
